webdataset prototype - RoutedDecoderIterableDataset

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50253 [WIP] webdataset prototype - class to preprocess image folder
* #49353 webdataset prototype - GroupByFilenameIterableDataset
* **#49140 webdataset prototype - RoutedDecoderIterableDataset**
* #49026 webdataset prototype - ReadFilesFromZipIterableDataset
* #49000 webdataset prototype - ReadFilesFromTarIterableDataset

Differential Revision: [D25554582](https://our.internmc.facebook.com/intern/diff/D25554582)

Commit:
webdataset prototype - RoutedDecoderIterableDataset

[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - RoutedDecoderIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - RoutedDecoderIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - RoutedDecoderIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - RoutedDecoderIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - RoutedDecoderIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - RoutedDecoderIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - RoutedDecoderIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - RoutedDecoderIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - RoutedDecoderIterableDataset"




[ghstack-poisoned]

Commit:
Update on "[WIP] webdataset prototype - RoutedDecoderIterableDataset"




[ghstack-poisoned]

Commit:
Update on "webdataset prototype - RoutedDecoderIterableDataset"




[ghstack-poisoned]

Commit:
Update on "webdataset prototype - RoutedDecoderIterableDataset"


Differential Revision: [D25554582](https://our.internmc.facebook.com/intern/diff/D25554582)

[ghstack-poisoned]

Commit:
Update on "webdataset prototype - RoutedDecoderIterableDataset"


Differential Revision: [D25554582](https://our.internmc.facebook.com/intern/diff/D25554582)

[ghstack-poisoned]

Commit:
Update on "webdataset prototype - RoutedDecoderIterableDataset"


Differential Revision: [D25554582](https://our.internmc.facebook.com/intern/diff/D25554582)

[ghstack-poisoned]

Commit:
Update on "webdataset prototype - RoutedDecoderIterableDataset"


Differential Revision: [D25554582](https://our.internmc.facebook.com/intern/diff/D25554582)

[ghstack-poisoned]

Commit:
Update on "webdataset prototype - RoutedDecoderIterableDataset"


Differential Revision: [D25554582](https://our.internmc.facebook.com/intern/diff/D25554582)

[ghstack-poisoned]

ejguan commented on torch/utils/data/datasets/decoder.py:
```diff
@@ -0,0 +1,270 @@
+        if isinstance(data, io.BufferedIOBase) or isinstance(data, io.RawIOBase):
+            data = data.read()
+
+        for f in self.handlers:
+            result = f(key, data)
+            if result is not None:
+                return result
```
Why we do iteration over all handlers? Several problems:
1. Hard to overwrite some extension handler.
2. Overhead for iteration over all handlers, and iteration over all extension types. (each data goes to all handlers until find corresponding)
3. Not find handler -> return None

glaringlee replied:
These codes are adopted from Nvidia, we can change it. 
1. They already have a helper function to overwrite extension handler (`handle_extension`).  
2. Iteration does have overhead, probably we can switch to dict or other types. 
3. We probably can raise error here.

ejguan commented on torch/utils/data/datasets/routeddecoderdataset.py:
```diff
@@ -0,0 +1,46 @@
+    def __iter__(self) -> Iterator[tuple]:
+        for data in self.dataset:
+            pathname = data[0]
+            result = self.decoder(data)
+            yield (pathname, result[pathname])
```
If result is None, pass to next.

glaringlee replied:
Probably raise.

ejguan commented on torch/utils/data/datasets/decoder.py:
```diff
@@ -0,0 +1,270 @@
+    def __init__(self, handlers):
+        self.handlers = handlers
+
+    def add_decoder(self, decoder):
+        if not decoder:
+            return
+        if not self.handlers:
+            self.handlers = [decoder]
+        else:
+            self.handlers.append(decoder)
```
It should be `add_handler` right?
`add_decoder` makes it hard to understand which one is decoder.

ejguan replied:
I propose two APIs:
1. add_handler
2. add_decoder

glaringlee replied:
Probably adding a comments or type annotation? Anything that is callable (with 0 arg) can be passed in as a decoder.

ejguan replied:
You mean 2 arguments, right? key and data

glaringlee replied:
Ah, yes, 2 arg.

ejguan commented on torch/utils/data/datasets/decoder.py:
```diff
@@ -0,0 +1,270 @@
+        return None
+
+    # add `type: ignore` to avoid mypy's warning on import missing
+    import torchvision.io  # type: ignore
```
Cycle dependency here.

glaringlee replied:
why this is cycle?

ejguan replied:
torchvision and torchaudio has dependency on torch, right?

glaringlee replied:
Yes, but not the other direction. 
torchvision.io/torchaudio import torch, but torch won't import torchvision.io/torchaudio, there is no cycle here to me, correct me if I am wrong. These code are from Nvidia, I might missed something here.

ejguan replied:
You are right, it's not circular dependency. Great to understand how python executes the import.
But, we should do `try import` here, since we don't explicitly have torchvision or torchaudio. I guess the webdataset has explicit requirements for these packages. (they remove it on October [link](https://github.com/tmbdev/webdataset/commit/f8cf56273d3324b86b22b24e47b2c7d72038de32#diff-4d7c51b1efe9043e44439a949dfd92e5827321b34082903477fd04876edb7552))
Besides, it's wield for me to import a package one time for every single data. For a huge dataset, the overhead can be crazy.

glaringlee replied:
Agree, this is not good generally. I think what Nvidia want to achieve here is that `if people don't need this handler, then don't import the needed library`, we can discuss further. I have an impression that python import is something like it check the module dict first, if no such module, then import it, the overhead here should be dict lookup? I am not 100% sure though. Let's revisit this after the tar dataset PR is landed.

ejguan commented on torch/utils/data/datasets/decoder.py:
```diff
@@ -0,0 +1,270 @@
+        return None
+
+    # add `type: ignore` to avoid mypy's warning on import missing
+    import torchaudio  # type: ignore
```
Cycle dependency

glaringlee replied:
same here, did I missed anything?

ejguan commented on torch/utils/data/datasets/routeddecoderdataset.py:
```diff
@@ -0,0 +1,46 @@
+            self,
+            dataset : Iterable,
+            *,
+            decoders : Union[None, List[str]] = None,
```
It seems the wrong type hint. Can we directly use `Decoder` as input?

glaringlee replied:
lol, why I put a str there :), should be callables.

Commit:
Update on "webdataset prototype - RoutedDecoderIterableDataset"


Differential Revision: [D25554582](https://our.internmc.facebook.com/intern/diff/D25554582)

[ghstack-poisoned]

Commit:
Update on "webdataset prototype - RoutedDecoderIterableDataset"


Differential Revision: [D25554582](https://our.internmc.facebook.com/intern/diff/D25554582)

[ghstack-poisoned]

Commit:
Update on "webdataset prototype - RoutedDecoderIterableDataset"


Differential Revision: [D25554582](https://our.internmc.facebook.com/intern/diff/D25554582)

[ghstack-poisoned]

Commit:
Update on "webdataset prototype - RoutedDecoderIterableDataset"


Differential Revision: [D25554582](https://our.internmc.facebook.com/intern/diff/D25554582)

[ghstack-poisoned]

glaringlee commented:
changed to datapipe and merged already
